### PR TITLE
Improve type checker performance

### PR DESCRIPTION
**Pull request by bschnurr**
# What does this PR do?

conditionally declare `TOKENIZER_MAPPING_NAMES` within a `if TYPE_CHECKING` block so that type checkers don't need to evaluate the RHS of the assignment.

this improves performance of the pylance/pyright type checkers

```Python
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("gpt2")

```

from 12 seconds, down to 2.5 seconds

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/auto/tokenization_auto.py [hunk 1]
```diff
@@ -43,153 +43,157 @@
...
 logger = logging.get_logger(__name__)
 
-
-TOKENIZER_MAPPING_NAMES = OrderedDict(
-    [
-        ("retribert", ("RetriBertTokenizer", "RetriBertTokenizerFast" if is_tokenizers_available() else None)),
-        ("roformer", ("RoFormerTokenizer", "RoFormerTokenizerFast" if is_tokenizers_available() else None)),
-        (
-            "t5",
+if TYPE_CHECKING:
```
Comment (sgugger) [inline src/transformers/models/auto/tokenization_auto.py:46]: `TYPE_CHECKING` needs to be imported to fix all the test errors.

File: src/transformers/models/auto/tokenization_auto.py [hunk 2]
```diff
@@ -43,153 +43,157 @@
...
+    # This significantly improves completion suggestion performance when
+    # the transformers package is used with Microsoft's Pylance language server.
+    TOKENIZER_MAPPING_NAMES: OrderedDict[str, Tuple[str, Optional[str]]] = OrderedDict()
```
Comment (sgugger) [inline src/transformers/models/auto/tokenization_auto.py:49]:
```suggestion
    TOKENIZER_MAPPING_NAMES: OrderedDict[str, Tuple[Optional[str], Optional[str]]] = OrderedDict()
```
First key is also optional if the slow tokenzier uses sentencepiece

**Review by sgugger — APPROVED** [review]
Thanks a lot for the PR! There are just a few things to fix before we can merge.

File: src/transformers/models/auto/tokenization_auto.py (hunk 1 above)
Comment (bschnurr) [reply src/transformers/models/auto/tokenization_auto.py:46]: oops. thanks!

**Comment by sgugger** [comment]
Last failure is not linked to this PR and has been fixed on master already, so we're good to go, thanks again!

**Pull request merged**